pkg/ccl/backupccl/backup_planning.go

```diff
@@ -1077,26 +1081,94 @@ func getScheduledBackupExecutionArgsFromSchedule(
 	return sj, args, nil
 }
 
+// writeBackupManifestCheckpoint writes a new BACKUP-CHECKPOINT MANIFEST
+// and CHECKSUM file. If it is a pure v22.1 cluster or later, it will write
+// a timestamped BACKUP-CHECKPOINT to the /progress directory.
+// If it is a mixed cluster version, it will write a non timestamped BACKUP-CHECKPOINT
+// to the base directory in order to not break backup jobs that resume
+// on a v21.2 node.
 func writeBackupManifestCheckpoint(
 	ctx context.Context,
-	jobID jobspb.JobID,
-	backupDetails jobspb.BackupDetails,
-	backupManifest *BackupManifest,
+	storageURI string,
+	encryption *jobspb.BackupEncryptionOptions,
+	desc *BackupManifest,
 	execCfg *sql.ExecutorConfig,
 	user security.SQLUsername,
 ) error {
-	defaultStore, err := execCfg.DistSQLSrv.ExternalStorageFromURI(ctx, backupDetails.URI, user)
+	defaultStore, err := execCfg.DistSQLSrv.ExternalStorageFromURI(ctx, storageURI, user)
 	if err != nil {
 		return err
 	}
 	defer defaultStore.Close()
 
-	if err := writeBackupManifest(
-		ctx, execCfg.Settings, defaultStore, tempCheckpointFileNameForJob(jobID),
-		backupDetails.EncryptionOptions, backupManifest,
-	); err != nil {
-		return errors.Wrapf(err, "writing checkpoint file")
+	sort.Sort(BackupFileDescriptors(desc.Files))
+
+	descBuf, err := protoutil.Marshal(desc)
+	if err != nil {
+		return err
+	}
+
+	descBuf, err = compressData(descBuf)
+	if err != nil {
+		return errors.Wrap(err, "compressing backup manifest")
 	}
+
+	if encryption != nil {
+		encryptionKey, err := getEncryptionKey(ctx, encryption, execCfg.Settings, defaultStore.ExternalIOConf())
+		if err != nil {
+			return err
+		}
+		descBuf, err = storageccl.EncryptFile(descBuf, encryptionKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We timestamp the checkpoint files in order to enforce write once
+	// backups. By timestamping with monotonically increasing timestamps
+	// we avoid ever overwriting a checkpoint file.
```

Comment:
nit: add a little comment around how we will consume these timestamped checkpoints

When the job goes to read these timestamped files, it will List the checkpoints and pick the file whose name is lexicographically sorted to the top. This will be the last checkpoint we write, for details refer to `newTimestampedCheckpointFileName`.